// azure-event-hubs/dist/lib/amqp-common/messageHeader.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const debugModule = require("debug");
const debug = debugModule("azure:amqp-common:messageHeader");
var MessageHeader;
(function (MessageHeader) {
    function toAmqpMessageHeader(props) {
        const amqpHeader = {
            delivery_count: props.deliveryCount,
            durable: props.durable,
            first_acquirer: props.firstAcquirer,
            priority: props.priority,
            ttl: props.ttl
        };
        debug("To AmqpMessageHeader: %O", amqpHeader);
        return amqpHeader;
    }
    MessageHeader.toAmqpMessageHeader = toAmqpMessageHeader;
    function fromAmqpMessageProperties(props) {
        const msgHeader = {
            deliveryCount: props.delivery_count,
            durable: props.durable,
            firstAcquirer: props.first_acquirer,
            priority: props.priority,
            ttl: props.ttl
        };
        debug("From AmqpMessageHeader: %O", msgHeader);
        return msgHeader;
    }
    MessageHeader.fromAmqpMessageProperties = fromAmqpMessageProperties;
})(MessageHeader = exports.MessageHeader || (exports.MessageHeader = {}));
