// azure-event-hubs/dist/lib/rhea-promise/receiver.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const debugModule = require("debug");
const _1 = require(".");
const debug = debugModule("rhea-promise:receiver");
class Receiver {
    constructor(session, receiver, options) {
        this._session = session;
        this._receiver = receiver;
        this.receiverOptions = options;
    }
    get name() {
        return this._receiver.name;
    }
    get error() {
        return this._receiver.error;
    }
    get properties() {
        return this._receiver.properties;
    }
    get source() {
        return this._receiver.source;
    }
    get target() {
        return this._receiver.target;
    }
    get address() {
        return this.source.address;
    }
    get session() {
        return this._session;
    }
    get connection() {
        return this._session.connection;
    }
    get drain() {
        return this._receiver.drain;
    }
    addCredit(credit) {
        this._receiver.add_credit(credit);
    }
    setCreditWindow(creditWindow) {
        this._receiver.set_credit_window(creditWindow);
    }
    /**
     * Determines whether the sender link is open.
     * @returns {boolean} `true` open. `false` closed.
     */
    isOpen() {
        let result = false;
        if (this._session.isOpen() && this._receiver.is_open()) {
            result = true;
        }
        return result;
    }
    remove() {
        if (this._receiver) {
            this._receiver.remove();
        }
        if (this._session) {
            this._session.remove();
        }
    }
    /**
     * Closes the amqp receiver.
     * @return {Promise<void>} Promise<void>
     * - **Resolves** the promise when rhea emits the "receiver_close" event.
     * - **Rejects** the promise with an AmqpError when rhea emits the
     * "receiver_error" event while trying to close an amqp receiver.
     */
    close() {
        const receiverClose = new Promise((resolve, reject) => {
            if (this.isOpen()) {
                let onError;
                let onClose;
                onClose = (context) => {
                    this._receiver.removeListener(_1.ReceiverEvents.receiverClose, onClose);
                    process.nextTick(() => {
                        debug("Resolving the promise as the amqp receiver has been closed.");
                        resolve();
                    });
                };
                onError = (context) => {
                    this._receiver.removeListener(_1.ReceiverEvents.receiverError, onError);
                    debug(`Error occurred while closing amqp receiver.`, context.session.error);
                    reject(context.session.error);
                };
                this._receiver.once(_1.ReceiverEvents.receiverClose, onClose);
                this._receiver.once(_1.ReceiverEvents.receiverError, onError);
                this._receiver.close();
            }
            else {
                resolve();
            }
        });
        return receiverClose.then(() => { return this._session.close(); });
    }
    registerHandler(event, handler) {
        this._receiver.on(event, handler);
    }
    removeHandler(event, handler) {
        this._receiver.removeListener(event, handler);
    }
    registerSessionHandler(event, handler) {
        this._session.registerHandler(event, handler);
    }
    removeSessionHandler(event, handler) {
        this._session.removeHandler(event, handler);
    }
}
exports.Receiver = Receiver;
